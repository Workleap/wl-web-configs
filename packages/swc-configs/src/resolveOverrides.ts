import type { Config } from "@swc/core";

export type ConfigOverrideFunction = (config: Config) => Config;

export type ConfigOverride = Config | ConfigOverrideFunction;

export function resolveOverrides(config: Config, configOverride?: ConfigOverride) {
    if (typeof configOverride === "function") {
        return {
            ...config,
            ...configOverride(config)
        };
    }

    if (configOverride) {
        return {
            ...config,
            ...configOverride
        };
    }

    return config;
}
