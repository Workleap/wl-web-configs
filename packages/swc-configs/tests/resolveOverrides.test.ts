import type { Config } from "@swc/core";
import { resolveOverrides } from "../src/resolveOverrides.ts";

test("when there are no config override, return the original config", () => {
    const config: Config = {
        jsc: {
            parser: {
                syntax: "typescript",
                tsx: true
            }
        }
    };

    const result = resolveOverrides(config);

    expect(result).toEqual(config);
});

describe("object", () => {
    test("when overrides are provided, merge the provided overrides with the config", () => {
        const config: Config = {
            jsc: {
                parser: {
                    syntax: "typescript",
                    tsx: true
                }
            }
        };

        const result = resolveOverrides(config, {
            module: {
                type: "es6"
            }
        });

        expect(result).toMatchSnapshot();
    });

    test("when an override match a config option, override the config option", () => {
        const config: Config = {
            jsc: {
                parser: {
                    syntax: "typescript",
                    tsx: true
                }
            }
        };

        const result = resolveOverrides(config, {
            jsc: {
                parser: {
                    syntax: "ecmascript"
                }
            }
        });

        expect(result).toMatchSnapshot();
    });
});

describe("function", () => {
    test("when overrides are provided, merge the provided overrides with the config", () => {
        const config: Config = {
            jsc: {
                parser: {
                    syntax: "typescript",
                    tsx: true
                }
            }
        };

        const result = resolveOverrides(config, () => ({
            module: {
                type: "es6"
            }
        }));

        expect(result).toMatchSnapshot();
    });

    test("when an override match a config option, override the config option", () => {
        const config: Config = {
            jsc: {
                parser: {
                    syntax: "typescript",
                    tsx: true
                }
            }
        };

        const result = resolveOverrides(config, () => ({
            jsc: {
                parser: {
                    syntax: "ecmascript"
                }
            }
        }));

        expect(result).toMatchSnapshot();
    });
});
