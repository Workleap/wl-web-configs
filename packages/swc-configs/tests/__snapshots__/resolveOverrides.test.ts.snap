// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function when an override match a config option, override the config option 1`] = `
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript",
    },
  },
}
`;

exports[`function when overrides are provided, merge the provided overrides with the config 1`] = `
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": true,
    },
  },
  "module": {
    "type": "es6",
  },
}
`;

exports[`object when an override match a config option, override the config option 1`] = `
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript",
    },
  },
}
`;

exports[`object when overrides are provided, merge the provided overrides with the config 1`] = `
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": true,
    },
  },
  "module": {
    "type": "es6",
  },
}
`;
